[VPlan] Introduce all loop regions as VPlan transform. (NFC)

fhahn opened the pull request:
Further simplify VPlan CFG builder by moving introduction of inner
regions to a VPlan transform, building on
https://github.com/llvm/llvm-project/pull/128419.

The HCFG builder now only constructs plain CFGs. I will move it to
VPlanConstruction as follow-up.

Depends on https://github.com/llvm/llvm-project/pull/128419, which is currently included in the PR as well.

fhahn committed:
[VPlan] Introduce child regions as VPlan transform.

Further simplify VPlan CFG builder by moving introduction of inner
regions to a VPlan transform, building on
https://github.com/llvm/llvm-project/pull/128419

The HCFG builder now only constructs plain CFGs. I will move it to
VPlanConstruction as follow-up.

iamlouk commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -0,0 +1,131 @@
+    if (HeaderVPBB->getNumPredecessors() != 2)
+      continue;
+    VPBlockBase *PreheaderVPBB = HeaderVPBB->getPredecessors()[0];
+    VPBlockBase *LatchVPBB = HeaderVPBB->getPredecessors()[1];
```
This would maybe be the chance to fix a long-standing bug in the VPlan-native path. LLVM will currently always canonicalize loops allowed in the VPlan-native path so that the latches second successor is the header in passes running before the vectorizer, but this is not guaranteed, and when relaxing restrictions on inner loops, then it does happen in the default LLVM pipeline that the loop exit condition is wrong for inner regions. The solution could look something like this:

```
if (LatchVPBB->getSuccessors()[1]  != HeaderVPBB) {
  auto *Term = LatchVPBB->getTerminator();
  auto *Not = new VPInstruction(VPInstruction::Not, {Term->getOperand(0)});
  Not->insertBefore(Term);
  Term->setOperand(0, Not);
}
```

fhahn replied:
Agreed that would be good, but probably best done separately. I added a test case to show the issue: 437d587e48002929552d982c9e2dc73428a37dd9

I'd be happy to share a follow-up fix, unless you'd be planning to do so?

fhahn replied:
I went ahead and put up a fix in https://github.com/llvm/llvm-project/pull/132292, so this doesn't get missed

fhahn reviewed (commented):
Rebased now that #128419 landed. Should be ready for review now.

fhahn commented:
ping :)

fhahn reviewed (commented):
Ping :)

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanHCFGBuilder.cpp:
```diff
@@ -376,23 +311,21 @@ void PlainCFGBuilder::buildPlainCFG(
     Loop *LoopForBB = LI->getLoopFor(BB);
     // Set VPBB predecessors in the same order as they are in the incoming BB.
     if (!isHeaderBB(BB, LoopForBB)) {
       setVPBBPredsFromBB(VPBB, BB);
```
Can setVPBBPredsFromBB() also handle header blocks?

fhahn replied:
For now, the code relies on the header having a specific predecessor order (preheader, then latch). We might be able to remove this restriction, because the dominance based approach could be generalized to handle it. It would require extra logic to determine which block is the preheader/latch.

fhahn replied:
I just pushed an update to not rely on the predecessor order (see cc818012a9683fd4d1fe4dceb3a5e2fb32d91a44 for this change in isolation)

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanHCFGBuilder.cpp:
```diff
@@ -376,23 +311,21 @@ void PlainCFGBuilder::buildPlainCFG(
 
     // Create VPInstructions for BB.
     createVPInstructionsForVPBB(VPBB, BB);
 
```
Can VPBB's successors be set according to BB's successors, in all cases, similar to predecessors. Or according to case order for switches, and true/false for conditional branches, but independent of loop latch or not?

fhahn replied:
Same as above, we could , with extra logic when introducing regions to figure out what successor is the exit block and which one the header.

Should be doable, but might be better separately?

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-all-regions-as-transform

fhahn committed:
!fixup address comments, thanks!

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-all-regions-as-transform

fhahn committed:
!fixup adjust names and comments after recent changes.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-all-regions-as-transform

fhahn committed:
!fixup don't special case header/latch predecessor order.

fhahn committed:
!fixup adjust comment

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanHCFGBuilder.cpp:
```diff
@@ -150,19 +118,6 @@ static bool isHeaderVPBB(VPBasicBlock *VPBB) {
-}
-
 // Create a new empty VPBasicBlock for an incoming BasicBlock in the region
 // corresponding to the containing loop  or retrieve an existing one if it was
```
Updated, thanks.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-all-regions-as-transform

fhahn committed:
!fixup address latest comments, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanHCFGBuilder.cpp:
```diff
@@ -113,13 +88,6 @@ static bool isHeaderBB(BasicBlock *BB, Loop *L) {
-
 // Add operands to VPInstructions representing phi nodes from the input IR.
 void PlainCFGBuilder::fixHeaderPhis() {
   for (auto *Phi : PhisToFix) {
```
Previosuly it wasnt consistent, but with the current patch it is; both use the order from the CFG in the current version and are a 1-1 translation of the CFG.

In the current patch, the canonicalization of header phi operand and header predecessor order is done completely separate from CFG construction, when the region is introduced.

In general, I think we should move away from directly relying on the operands to retrieve incoming values and move towards an interface `getIncomingValueFor(VPB)`, which takes care of checking what the index for `VPB` is.

ayalz replied:
> Responding to some comments inline regarding phi/predecessor ordering to start with, to clarify. Would be good to decide first where we want the canonicalization of phi/predecessors to happen.
> 
> Doing it separately from initial VPlan construction has the benefit of reducing the number of special cases needed during initial construction and shifts the complexity of detecting the preheader/latch to a VPlan transform, currently integrated with introducing regions, but could also be a separate canonicalization.

Ah, ok, I see. Sketching how separate canonicalization may look like, for clarify and to help reason about what if anything should be fused:

Let VPlanCFG0 denote the initial VPlan, constructed from IR. VPlanCFG0 may indeed be a consistent 1-1 translation of the CFG and IR, arguably with all its branches - including early-exits and latch-to-exit (and perhaps unconditional branches too), possibly also representing the BasicBlock operands of phi's and branches as VPValues wrapping these Values, as operands of the corresponding VPIRInstructions.

Let VPlanCFG1 denote the result of canonicalizing VPlanCFG0's phi's and branches according to predecessors/successors order (instead of their BasicBlock operands). It may be better to change the types of recipes used, even when the original predecessor/successor order is retained, to clarify that they belong to distinct VPlan "modes" or "stages". Early-exit branch recipes and successors can be eliminated now.

Let VPlanHCFG0 denote the (first) HCFG VPlan produced by a VPlan transform devoted to canonical-CFG --> HCFG conversion that introduces child (or actually all loop) regions into VPlanCFG1. Would this transform need to update/change any recipes?

fhahn replied:
> Let VPlanCFG0 denote the initial VPlan, constructed from IR. VPlanCFG0 may indeed be a consistent 1-1 translation of the CFG and IR, arguably with all its branches - including early-exits and latch-to-exit (and perhaps unconditional branches too), possibly also representing the BasicBlock operands of phi's and branches as VPValues wrapping these Values, as operands of the corresponding VPIRInstructions.


Yep, I have some follow-up patches to retain the exit branches and keep the exit blocks connected.

In a sense, the specific operand order (1st op = preheader, 2nd op = latch) for header phis is purely a consequence of introducing a region, which itself only has a single predecessor I think. So introducing the canonical order at this point seems like a good fit; the header phis are the only ones which need their order adjusted.

Using a separate recipe for the initial recipes sounds good, something like VPInstruction which always has an underlying instruction would be good, but already models the operands/defs in VPlan via operands. Unfortunately VPIRInstruction is already taken. We would need a version that defines result VPValues and uses VPvalue as operands.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -14,26 +14,93 @@
+/// header block, returns a pair with the corresponding preheader and latch
+/// blocks. Otherwise return std::nullopt.
+static std::optional<std::pair<VPBasicBlock *, VPBasicBlock *>>
+getPreheaderAndLatch(VPBlockBase *HeaderVPB, const VPDominatorTree &VPDT) {
```
This means effectively going back to the orignial verison, which enforced the specific order? The current version moves the canonicalization of order outside the CFG construction, and leaves the plain CFG a 1-1 mapping of the incomig CFG, with both phi recipes and VPBB predecessor order matching the order in the original CFG.

fhahn reviewed (commented):
Responding to some comments inline regarding phi/predecessor ordering to start with, to clarify. Would be good to decide first where we want the canonicalization of phi/predecessors to happen.

Doing it separately from initial VPlan construction has the benefit of reducing the number of special cases needed during initial construction and shifts the complexity of detecting the preheader/latch to a VPlan transform, currently integrated with introducing regions, but could also be a separate canonicalization.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-all-regions-as-transform

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9312,14 +9312,14 @@ LoopVectorizationPlanner::tryToBuildVPlanWithVPRecipes(VFRange &Range) {
   auto Plan = std::make_unique<VPlan>(OrigLoop);
   // Build hierarchical CFG.
-  // Convert to VPlan-transform and consoliate all transforms for VPlan
+  // TODO: Convert to VPlan-transform and consoliate all transforms for VPlan
```
Done thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -14,26 +14,93 @@
+  VPDominatorTree VPDT;
+  VPDT.recalculate(Plan);
+  for (VPBlockBase *HeaderVPB : vp_depth_first_shallow(Plan.getEntry()))
+    createLoopRegion(Plan, HeaderVPB, VPDT);
```
Updated the name to `tryToCreateLoopRegion`, to avoid adding new arguments, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -14,26 +14,57 @@
-         "Plan should end at top level latch");
+  auto *HeaderVPBB = cast<VPBasicBlock>(Plan.getEntry()->getSingleSuccessor());
+  VPRegionBlock *TopRegion = introduceRegion(Plan, HeaderVPBB, VPDT);
+  auto *OrigExiting = TopRegion->getExiting();
+  VPBasicBlock *LatchVPBB = Plan.createVPBasicBlock("vector.latch");
+  VPBlockUtils::insertBlockAfter(LatchVPBB, OrigExiting);
+  TopRegion->setExiting(LatchVPBB);
```
Done, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanHCFGBuilder.cpp:
```diff
@@ -130,43 +98,23 @@ void PlainCFGBuilder::fixHeaderPhis() {
-    assert(isHeaderBB(Phi->getParent(), L));
+    assert(isHeaderBB(Phi->getParent(), LI->getLoopFor(Phi->getParent())));
     // For header phis, make sure the incoming value from the loop
     // predecessor is the first operand of the recipe.
```
I dropped the comment, I think at this point the header phis are not really special (the fixed predecessor order comes when we introduce the corresponding regions)

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanHCFGBuilder.cpp:
```diff
@@ -130,43 +98,23 @@ void PlainCFGBuilder::fixHeaderPhis() {
 
 static bool isHeaderVPBB(VPBasicBlock *VPBB) {
   return VPBB->getParent() && VPBB->getParent()->getEntry() == VPBB;
 }
```
Removed, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanHCFGBuilder.cpp:
```diff
@@ -423,21 +338,11 @@ void PlainCFGBuilder::buildPlainCFG(
+    // Don't connect any blocks outside the current loop except the latches for
+    // inner loops.
+    if (LoopForBB &&
+        (LoopForBB == TheLoop || BB != LoopForBB->getLoopLatch())) {
```
> This is somewhat confusing: can LoopForBB be null, given that BB traverses all (and only) blocks of TheLoop?

It cannot be null, I removed the check.

Initial VPlan-construction should also include connecting the exit blocks of the top-level loop eventually. I have a follow-up patch to do so, as it is probably best tackled separately.

fhahn committed:
!fixup adderss latest comments, thanks!

fhahn reviewed (commented):
I left the code dealing with adding operands to phis as-is (keep IR order for both predecessors and operands during intiial construction, canonicalize when region is introduced), pending adjustment if needed.

All other comments should be addressed. Ping :)

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -14,26 +14,95 @@
-  VPBlockUtils::connectBlocks(Plan.getEntry(), VecPreheader);
-  assert(OriginalLatch->getNumSuccessors() == 0 &&
-         "Plan should end at top level latch");
+  auto *PreheaderVPBB = cast<VPBasicBlock>(Preds[0]);
+  auto *LatchVPBB = cast<VPBasicBlock>(Preds[1]);
```
```suggestion
  auto *PreheaderVPBB = Preds[0];
  auto *LatchVPBB = Preds[1];
```
or are the casts needed?

fhahn replied:
Removed, thanks

ayalz reviewed (commented):
> I left the code dealing with adding operands to phis as-is (keep IR order for both predecessors and operands during intiial construction, canonicalize when region is introduced), pending adjustment if needed.

OK. Would it be better to apply this canonicalization when looking for region headers - swapping the predecessors of the header together with the operands of its phi recipes? Admittedly these predecessors are soon to be disconnected, but the number of headers (that require clear yet somewhat redundant swapping) is small.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-all-regions-as-transform

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -14,26 +14,95 @@
+      VPDT.dominates(HeaderVPB, LatchVPBB))
+    return {std::make_pair(PreheaderVPBB, LatchVPBB)};
+
+  return std::nullopt;
```
Done, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -14,26 +14,95 @@
+  std::swap(PreheaderVPBB, LatchVPBB);
+  if (VPDT.dominates(PreheaderVPBB, HeaderVPB) &&
+      VPDT.dominates(HeaderVPB, LatchVPBB))
+    return {std::make_pair(PreheaderVPBB, LatchVPBB)};
```
Done thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -14,26 +14,95 @@
+  auto *LatchVPBB = cast<VPBasicBlock>(Preds[1]);
+  if (VPDT.dominates(PreheaderVPBB, HeaderVPB) &&
+      VPDT.dominates(HeaderVPB, LatchVPBB))
+    return {std::make_pair(PreheaderVPBB, LatchVPBB)};
```
Done thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -14,26 +14,95 @@
+getPreheaderAndLatch(VPBlockBase *HeaderVPB, const VPDominatorTree &VPDT) {
+  ArrayRef<VPBlockBase *> Preds = HeaderVPB->getPredecessors();
+  if (Preds.size() != 2)
+    return std::nullopt;
```
done thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -14,26 +14,95 @@
+/// header block, returns a pair with the corresponding preheader and latch
+/// blocks. Otherwise return std::nullopt.
+static std::optional<std::pair<VPBasicBlock *, VPBasicBlock *>>
+getPreheaderAndLatch(VPBlockBase *HeaderVPB, const VPDominatorTree &VPDT) {
```
Done thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -14,26 +14,95 @@
+      R.setOperand(0, R.getOperand(1));
+      R.setOperand(1, Inc0);
+    }
+  }
```
done thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -14,26 +14,95 @@
+  VPDominatorTree VPDT;
+  VPDT.recalculate(Plan);
+  for (VPBlockBase *HeaderVPB : vp_depth_first_shallow(Plan.getEntry()))
+    tryToCreateLoopRegion(Plan, HeaderVPB, VPDT);
```
done thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanHCFGBuilder.cpp:
```diff
@@ -130,43 +98,23 @@ void PlainCFGBuilder::fixHeaderPhis() {
-
-    Loop *L = LI->getLoopFor(Phi->getParent());
-    assert(isHeaderBB(Phi->getParent(), L));
+    assert(isHeaderBB(Phi->getParent(), LI->getLoopFor(Phi->getParent())));
```
Done thanks

fhahn committed:
!fixup address latest comments, thanks

fhahn commented:
> > I left the code dealing with adding operands to phis as-is (keep IR order for both predecessors and operands during intiial construction, canonicalize when region is introduced), pending adjustment if needed.
> 
> OK. Would it be better to apply this canonicalization when looking for region headers - swapping the predecessors of the header together with the operands of its phi recipes? Admittedly these predecessors are soon to be disconnected, but the number of headers (that require clear yet somewhat redundant swapping) is small.

Sounds good, updated, thanks!

ayalz reviewed (approved):
LGTM, thanks!

ayalz commented:
"Introduce all loop regions as VPlan transform" may be a more accurate title than "Introduce child regions as VPlan transform."

fhahn commented:
> "Introduce all loop regions as VPlan transform" may be a more accurate title than "Introduce child regions as VPlan transform."

Thanks, updated! Also added NFC, as it only changes minor debug output
